tests/xdsl_opt/test_xdsl_opt.py

```diff
@@ -44,57 +44,83 @@ def test_empty_program():
 
 
 @pytest.mark.parametrize(
-    "args, expected_error",
+    "args, error_type, expected_error",
     [
         (
             ["--no-implicit-module", "tests/xdsl_opt/not_module.mlir"],
+            ParseError,
             "builtin.module operation expected",
         ),
         (
             ["--no-implicit-module", "tests/xdsl_opt/incomplete_program.mlir"],
+            ParseError,
             "Could not parse entire input",
         ),
         (
             ["tests/xdsl_opt/incomplete_program_residual.mlir"],
+            ParseError,
             "Could not parse entire input",
         ),
         (
             ["tests/xdsl_opt/incomplete_program.mlir"],
+            ParseError,
             "Could not parse entire input",
         ),
-        (["tests/xdsl_opt/empty_program.wrong"], "Unrecognized file extension 'wrong'"),
+        (
+            ["tests/xdsl_opt/empty_program.wrong"],
+            ValueError,
+            "Unrecognized file extension 'wrong'",
+        ),
+        (
+            ["tests/xdsl_opt/unverified_program.mlir"],
+            DiagnosticException,
+            "operand at position 0 does not verify",
+        ),
     ],
 )
-def test_error_on_run(args: list[str], expected_error: str):
+def test_error_on_run(
+    args: list[str], error_type: type[Exception], expected_error: str
+):
     opt = xDSLOptMain(args=args)
 
-    with pytest.raises(Exception, match=expected_error):
+    with pytest.raises(error_type, match=expected_error):
         opt.run()
 
 
 @pytest.mark.parametrize(
-    "args, expected_error",
+    "args, error_type, expected_error",
     [
         (
             ["tests/xdsl_opt/empty_program.mlir", "-p", "wrong"],
+            Exception,
             "Unrecognized passes: ['wrong']",
-        )
+        ),
     ],
 )
-def test_error_on_construction(args: list[str], expected_error: str):
-    with pytest.raises(Exception, match=re.escape(expected_error)):
-        _opt = xDSLOptMain(args=args)
-
-
-def test_wrong_target():
-    filename = "tests/xdsl_opt/empty_program.mlir"
-    opt = xDSLOptMain(args=[filename])
-    opt.args.target = "wrong"
+def test_error_on_construction(
+    args: list[str], error_type: type[Exception], expected_error: str
+):
+    with pytest.raises(error_type, match=re.escape(expected_error)):
+        xDSLOptMain(args=args)
 
-    with pytest.raises(Exception) as e:
-        opt.run()
 
-    assert e.value.args[0] == "Unknown target wrong"
+@pytest.mark.parametrize(
+    "args, expected_error",
+    [
+        (
+            ["tests/xdsl_opt/empty_program.mlir", "-t", "wrong"],
+            "invalid choice: 'wrong'",
+        ),
+    ],
+)
+def test_error_on_argparse(
+    capsys: pytest.CaptureFixture[str], args: list[str], expected_error: str
```

Comment:
Found out about this helper while googling for how to test argparse, we might want to use it in other tests in the future

Reply:
Yep exactly, it captures stdout and stderr